jcl/src/java.base/share/classes/java/lang/invoke/MethodHandleHelper.java

```diff
@@ -206,7 +206,54 @@ private static final Object invokeBsm(MethodHandle bsm, Object[] staticArgs) thr
 	}
 	
 	@SuppressWarnings("unused")
-	private static final MethodHandle resolveInvokeDynamic(long j9class, String name, String methodDescriptor, long bsmData) throws Throwable {
+	private static final Object resolveInvokeDynamic(long j9class, String name, String methodDescriptor, long bsmData) throws Throwable {
+/*[IF OPENJDK_METHODHANDLES]*/
+		Object[] result = new Object[2];
+		MethodType type = null;
+
+		try {
+			VMLangAccess access = VM.getVMLangAccess();
+			Object internalRamClass = access.createInternalRamClass(j9class);
+			Class<?> classObject = getClassFromJ9Class(j9class);
+			
+			type = MethodType.vmResolveFromMethodDescriptorString(methodDescriptor, access.getClassloader(classObject), null);
+
+			int bsmIndex = UNSAFE.getShort(bsmData);
+			int bsmArgCount = UNSAFE.getShort(bsmData + BSM_ARGUMENT_COUNT_OFFSET);
+			long bsmArgs = bsmData + BSM_ARGUMENTS_OFFSET;
+			MethodHandle bsm = getCPMethodHandleAt(internalRamClass, bsmIndex);
+			if (null == bsm) {
+				/*[MSG "K05cd", "unable to resolve 'bootstrap_method_ref' in '{0}' at index {1}"]*/
+				throw new NullPointerException(Msg.getString("K05cd", classObject.toString(), bsmIndex)); //$NON-NLS-1$
+			}
+			Object[] staticArgs = new Object[bsmArgCount];
+
+			/* Static optional arguments */
+			int bsmTypeArgCount = bsm.type().parameterCount();
+			for (int i = 0; i < bsmArgCount; i++) {
+				staticArgs[i] = getAdditionalBsmArg(access, internalRamClass, classObject, bsm, bsmArgs, bsmTypeArgCount, i);
+			}
+
+			Object[] appendixResult = new Object[1];
+			appendixResult[0] = null;
+			try {
+				MemberName mname = MethodHandleNatives.linkCallSite(classObject, 0, bsm, name, type, (Object)staticArgs, appendixResult);
+
+				result[0] = mname;
+				result[1] = appendixResult[0];
+			} catch (Throwable e) {
+				if (e instanceof Error) {
+					result[0] = e;
+				} else {
+					result[0] = new BootstrapMethodError(e);
+				}
+			}
+		} catch (Throwable e) {
+			result[0] = e;
```

Comment:
Which PR has the code that does the `invokedynamic`?  I need to review that and this code at the same time to have any chance of telling if it's correct or not

Reply:
https://github.com/eclipse/openj9/pull/10762: Has the changes for resolvesupport.cpp, callin.cpp, ...
https://github.com/eclipse/openj9/pull/10733: Has the BytecodeInterpreter changes for invokedynamic

Reply:
Looking at the bytecode changes, this requires a type check on every invokedynamic in the interpreter which I assume the JIT can avoid by checking at compile time:
https://github.com/eclipse/openj9/blob/2e5ef9cf8a015cc34f59e10516c69b97bdfc3364/runtime/vm/BytecodeInterpreter.hpp#L8022-L8025

With the original J9 implementation, we resolved to a MH that dropped the incoming args and then threw the exception.  Can we not play the same trick here with the OJDK implementation?

Reply:
@DanHeidinga The OJDK implementation generates adapter method (memberName) to invoke a method handle.
If we want to use similar trick as current OpenJ9 impl, then we need to find a way to generate a MN that points to a method which accepts vararg input and throw last argument (ie. we place the exception object in the appendix slot).

We will have to investigate how to create such memberName in java.
@babsingh do you have further insight on this?

Reply:
No new insights. We should investigate/find how to `generate a MN that points to a method which accepts vararg input and throw last argument (i.e. we place the exception object in the appendix slot)`. Adding this as a TODO task for the future once this PR is merged. I will also leave a comment in the code.

Reply:
Added a TODO comment for the above task.